src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1948,10 +1948,23 @@ private void CheckMemberNameConflicts(BindingDiagnosticBag diagnostics)
                         {
                             if (symbol.Kind != SymbolKind.Field || !symbol.IsImplicitlyDeclared)
                             {
-                                // The type '{0}' already contains a definition for '{1}'
-                                if (Locations.Length == 1 || IsPartial)
+                                var explicitImplementations = symbol.GetExplicitInterfaceImplementations();
+                                var lastExplicitImplementations = lastSym.GetExplicitInterfaceImplementations();
+
+                                // Source symbols can only explicitly implement at most 1 member
+                                Debug.Assert(explicitImplementations.Length is 0 or 1 && explicitImplementations.Length == lastExplicitImplementations.Length);
+
+                                // Sometimes two non-overloadable members can have exactly the same source name, but be explicit implementations of different members.
+                                // For example, when explicitly implementing members of file-local interfaces.
+                                // We don't consider such members to be duplicates.
+                                if (explicitImplementations.Length == 0
+                                    || explicitImplementations[0].Equals(lastExplicitImplementations[0], TypeCompareKind.AllIgnoreOptions))
```

Comment:
This looks like a "voodoo magic" to me. It doesn't look appropriate to suppress the error based on the logic.

Reply:
Note that the similar situation (when names in source match) can be created with regular types. In this case, the names for explicit interface implementations are made by compiler unique:
```
    [Fact, WorkItem("[https://github.com/dotnet/roslyn/issues/68219](https://github.com/dotnet/roslyn/issues/68219)")]
    public void PartialExplicitImplementation_05_()
    {
        var source0 = """
            var c = new C();
            c.Use1();
            c.Use2();
            """;

        var source1 = """
            using System;
            using NS1;
            
            namespace NS1
            {
                interface FI
                {
                    int Bar { get; }
                }
            }

            internal partial class C : FI
            {
                int FI.Bar => 1;
                public void Use1() => Console.Write(((FI)this).Bar);
            }
            """;

        var source2 = """
            using System;
            using NS2;

            namespace NS2
            {
                interface FI
                {
                    int Bar { get; }
                }
            }
            
            internal partial class C : FI
            {
                int FI.Bar => 2;
                public void Use2() => Console.Write(((FI)this).Bar);
            }
            """;

        var verifier = CompileAndVerify(new[] { (source0, "F0.cs"), (source1, "F1.cs"), (source2, "F2.cs") }, expectedOutput: "12");
        verifier.VerifyDiagnostics();
        verifier.VerifyTypeIL("C",
@"
.class private auto ansi beforefieldinit C
 extends [mscorlib]System.Object
 implements NS1.FI,
            NS2.FI
{
 // Methods
 .method private final hidebysig specialname newslot virtual 
  instance int32 NS1.FI.get_Bar () cil managed 
 {
  .override method instance int32 NS1.FI::get_Bar()
  // Method begins at RVA 0x2081
  // Code size 2 (0x2)
  .maxstack 8
  IL_0000: ldc.i4.1
  IL_0001: ret
 } // end of method C::NS1.FI.get_Bar
 .method public hidebysig 
  instance void Use1 () cil managed 
 {
  // Method begins at RVA 0x2084
  // Code size 12 (0xc)
  .maxstack 8
  IL_0000: ldarg.0
  IL_0001: callvirt instance int32 NS1.FI::get_Bar()
  IL_0006: call void [mscorlib]System.Console::Write(int32)
  IL_000b: ret
 } // end of method C::Use1
 .method private final hidebysig specialname newslot virtual 
  instance int32 NS2.FI.get_Bar () cil managed 
 {
  .override method instance int32 NS2.FI::get_Bar()
  // Method begins at RVA 0x2091
  // Code size 2 (0x2)
  .maxstack 8
  IL_0000: ldc.i4.2
  IL_0001: ret
 } // end of method C::NS2.FI.get_Bar
 .method public hidebysig 
  instance void Use2 () cil managed 
 {
  // Method begins at RVA 0x2094
  // Code size 12 (0xc)
  .maxstack 8
  IL_0000: ldarg.0
  IL_0001: callvirt instance int32 NS2.FI::get_Bar()
  IL_0006: call void [mscorlib]System.Console::Write(int32)
  IL_000b: ret
 } // end of method C::Use2
 .method public hidebysig specialname rtspecialname 
  instance void .ctor () cil managed 
 {
  // Method begins at RVA 0x2079
  // Code size 7 (0x7)
  .maxstack 8
  IL_0000: ldarg.0
  IL_0001: call instance void [mscorlib]System.Object::.ctor()
  IL_0006: ret
 } // end of method C::.ctor
 // Properties
 .property instance int32 NS1.FI.Bar()
 {
  .get instance int32 C::NS1.FI.get_Bar()
 }
 .property instance int32 NS2.FI.Bar()
 {
  .get instance int32 C::NS2.FI.get_Bar()
 }
} // end of class C
");
    }
```

I think we simply should do the "same" thing for file-local types situation.

Reply:
For an explicit interface implementation, should we adjust the member's name when the symbol is initially created to include the "file-local" prefix?

Reply:
> For an explicit interface implementation, should we adjust the member's name when the symbol is initially created to include the "file-local" prefix?

Probably. I would do that in the same place we "mangle" the names for the test scenario above.